transport/gdrive-transport/src/main/java/org/apache/airavata/mft/transport/gdrive/GDriveSender.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.airavata.mft.secret.service.GDriveSecret;
+import org.apache.airavata.mft.secret.service.GDriveSecretGetRequest;
+import org.apache.airavata.mft.secret.service.SecretServiceGrpc;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+
+
+public class GDriveSender implements Connector {
+    private static final Logger logger = LoggerFactory.getLogger(GDriveSender.class);
+
+    private GDriveResource gdriveResource;
+    private Drive drive;
+    private JsonObject jsonObject;
+
+
+    @Override
+    public void init(String resourceId, String credentialToken, String resourceServiceHost, int resourceServicePort, String secretServiceHost, int secretServicePort) throws Exception {
+        ResourceServiceGrpc.ResourceServiceBlockingStub resourceClient = ResourceServiceClient.buildClient(resourceServiceHost, resourceServicePort);
+        this.gdriveResource = resourceClient.getGDriveResource(GDriveResourceGetRequest.newBuilder().setResourceId(resourceId).build());
+
+        SecretServiceGrpc.SecretServiceBlockingStub secretClient = SecretServiceClient.buildClient(secretServiceHost, secretServicePort);
+        GDriveSecret gdriveSecret = secretClient.getGDriveSecret(GDriveSecretGetRequest.newBuilder().setSecretId(credentialToken).build());
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        String jsonString = gdriveSecret.getCredentialsJson();
+        jsonObject = new JsonParser().parse(jsonString).getAsJsonObject();
+        GoogleCredential credential = GoogleCredential.fromStream(new ByteArrayInputStream(jsonString.getBytes(StandardCharsets.UTF_8)), transport, jsonFactory);
+
+        if (credential.createScopedRequired()) {
+            Collection<String> scopes = DriveScopes.all();
+            credential = credential.createScoped(scopes);
+        }
+
+        drive = new Drive.Builder(transport, jsonFactory, credential).build();
+
+    }
+
+    @Override
+    public void destroy() {
+
+    }
+
+    @Override
+    public void startStream(ConnectorContext context) throws Exception {
+        logger.info("Starting GDrive send for remote server for transfer {} ", context.getTransferId());
+        String id = null;
+        boolean fileupdated = false;
+        String entityUser = jsonObject.get("client_email").getAsString();
+        File fileMetadata = new File();
+
+        InputStreamContent contentStream = new InputStreamContent(
+                "", context.getStreamBuffer().getInputStream());
+
+
+        fileMetadata.setName(this.gdriveResource.getResourcePath());
+        FileList fileList = drive.files().list()
+                .setQ("name = '"+gdriveResource.getResourcePath()+"'")
+                .setFields("files(id,name)")
+                .execute();
+
+        for (File f : fileList.getFiles()) {
```

Comment:
validate fileList.getFiles() as I have mentioned before

Reply:
Done for empty filelist as above

Reply:
Where did you do that validation? I can't see in the latest code

Reply:
you should do something like below and get rid of other sections. This makes code more readable and error safe

Files[] filtered = fileList.getFiles()
if (filtered == null || filtered.length == 0) {
     File file = drive.files().create(fileMetadata, contentStream).setFields("id").execute();
     Permission userPermission = new Permission();
     userPermission.setType("user").setRole("writer").setEmailAddress(entityUser);
     drive.permissions().create(file.getId(), userPermission).execute();
} else if (filtered.length > 1) {
     throw new Exception("More than one file found with name " + gdriveResource.getResourcePath());
} else {
    id = filtered[0].getId();
    File file = drive.files().get(id).execute();
    drive.files().update(file.getId(), fileMetadata, contentStream).setFields("id").execute();
}

Reply:
Changes added and tested